rfc/0001-single_memory_mapped_file.md

```diff
@@ -0,0 +1,70 @@
+# Name
+[name]: #name
+
+Single Memory-Mapped File
+
+# Author
+[author]: #author
+
+Dragan Milic
+
+# Summary
+[summary]: #summary
+
+Using a memory mapped file to store the database data.
+Instead of using file read and write operations, we should be using a memory-mapped file.
+
+# Motivation
+[motivation]: #motivation
+
+Usual syscalls to read/write data (such as _open()_/_read()_/_write()_/...) are designed for sequential processing of files.
+In the case of a database, this is sub-optimal, since we need a random access to the file data.
+
+# Design
+[design]: #design
+
+Data of the whole database should be stored in one file.
+When the database is opened, this file is memory-mapped (using the _mmap()_ syscall) to a memory region.
+After the file is memory-mapped, data can be accessed via memory access (in Golang using a byte slice).
+Writing data to the file is as simple as copying data to that memory region.
+
+## Advantages
+
+Memory mapped files can be much larger than the available RAM, but can be accessed as if they were just sequential portion of memory.
+This is achieved by leverages OS's paging mechanism, loading pages on access and keeping them paged until there is a need to free memory for other purposes.
+
+One of the advantages is that mmap-ed files won't be handled by Golang's garbage collector and the overhead of a context switch will be only incurred if the page is not yet in RAM.
+In addition, the task of caching of pages kept in memory are completely outsourced to the OS running the application, enabling it to use all available memory of the system if needed.
+
+## Appending data
+
+When a file is memory-mapped, one of the parameters is the size of the memory region for the map.
+This size can be larger than the size of the file.
+If the memory region is larger than the current file length, when the file grows, additional data is automatically available in the memory-mapped region.
+
+We can use that to append data to existing database - we can just extend the file using _ftruncate()_ syscall.
```

Comment:
is `ftruncate()` as POSIX call or only available on Linux?

Reply:
yes, it is